lib/SILOptimizer/IPO/GlobalOpt.cpp

```diff
@@ -966,27 +1090,104 @@ bool SILGlobalOpt::run() {
           continue;
         }
 
-        auto *GAI = dyn_cast<GlobalAddrInst>(&I);
-        if (!GAI) {
+        if (auto *GAI = dyn_cast<GlobalAddrInst>(&I)) {
+          collectGlobalAccess(GAI);
           continue;
         }
 
-        collectGlobalAccess(GAI);
+        if (auto *allocGlobal = dyn_cast<AllocGlobalInst>(&I)) {
+          collectAllocGlobal(allocGlobal->getReferencedGlobal(), allocGlobal);
+          continue;
+        }
       }
     }
   }
+}
+
+bool SILGlobalOpt::run() {
+  // Collect all the global variables and associated instructions.
+  collect();
 
+  // Optimize based on what we just collected.
   for (auto &InitCalls : GlobalInitCallMap) {
+    // Don't optimize functions that are marked with the opt.never attribute.
+    bool shouldOptimize = true;
+    for (auto *apply : InitCalls.second) {
+      if (!apply->getFunction()->shouldOptimize()) {
+        shouldOptimize = false;
+        break;
+      }
+    }
+    if (!shouldOptimize)
+      continue;
+
     // Optimize the addressors if possible.
     optimizeInitializer(InitCalls.first, InitCalls.second);
     placeInitializers(InitCalls.first, InitCalls.second);
   }
 
   for (auto &Init : GlobalVarStore) {
+    // Don't optimize functions that are marked with the opt.never attribute.
+    if (!Init.second->getFunction()->shouldOptimize())
+      continue;
+
     // Optimize the access to globals if possible.
     optimizeGlobalAccess(Init.first, Init.second);
   }
 
+  SmallVector<SILGlobalVariable *, 8> addrGlobals;
+  for (auto &addrPair : GlobalAddrMap) {
+    // Don't optimize functions that are marked with the opt.never attribute.
+    bool shouldOptimize = true;
+    for (auto *addr : addrPair.second) {
+      if (!addr->getFunction()->shouldOptimize()) {
+        shouldOptimize = false;
+        break;
+      }
+    }
+    if (!shouldOptimize)
+      continue;
+
+    addrGlobals.push_back(addrPair.first);
+  }
+
+  for (auto *global : addrGlobals) {
+    HasChanged |= tryRemoveGlobalAddr(global);
+  }
+
+  SmallVector<std::pair<SILGlobalVariable *, AllocGlobalInst *>, 12>
+      globalAllocPairs;
+  for (auto &alloc : AllocGlobalStore) {
+    if (!alloc.second->getFunction()->shouldOptimize())
+      continue;
+    globalAllocPairs.push_back(std::make_pair(alloc.first, alloc.second));
+  }
+
+  for (auto &allocPair : globalAllocPairs) {
+    HasChanged |= tryRemoveGlobalAlloc(allocPair.first, allocPair.second);
+  }
+
+  // Erase the instructions that we have marked for deletion.
+  for (auto *inst : InstToRemove) {
+    inst->eraseFromParent();
+  }
+
+  // After we erase some instructions, re-collect.
+  reset();
+  collect();
+
+  // Copy the globals so we don't get issues with modifying while iterating.
+  SmallVector<SILGlobalVariable *, 12> globals;
+  for (auto &global : Module->getSILGlobals()) {
+    globals.push_back(&global);
```

Comment:
This shouldn't be necessary if tryRemoveUnusedGlobal() instead adds the globals to remove onto a new list.

Reply:
Good catch.